pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,368 @@
+		log.Error(err, "More than one revision exists for the given templateHash")
+		return revisions[len(revisions)-1], nil
+	}
+
+	return revisions[0], nil
+}
+
+// GetPatch returns a strategic merge patch that can be applied to restore a LeaderWorkerSet to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is the
+// leaderWorkerTemplate and NetworkConfig. We can modify this later to encompass more state (or less) and
+// remain compatible with previously recorded patches.
+func GetPatch(lws *leaderworkerset.LeaderWorkerSet) ([]byte, error) {
+	str := &bytes.Buffer{}
+	clone := lws.DeepCopy()
+	// When upgrading from an LWS version that doesn't contain NetworkConfig, NetworkConfig will be nil
+	// until another field in the LWS object is changed triggering the LWS webhook. This allows the revision
+	// to be the same before and after the LWS webhook actually defaults the value.
+	if clone.Spec.NetworkConfig == nil {
+		clone.Spec.NetworkConfig = &leaderworkerset.NetworkConfig{}
+		subdomainPolicy := leaderworkerset.SubdomainShared
+		clone.Spec.NetworkConfig = &leaderworkerset.NetworkConfig{
+			SubdomainPolicy: &subdomainPolicy,
+		}
+	}
+
+	if err := unstructured.UnstructuredJSONScheme.Encode(clone, str); err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	if err := json.Unmarshal(str.Bytes(), &raw); err != nil {
+		return nil, err
+	}
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
+	specCopy["networkConfig"] = spec["networkConfig"]
+	specCopy["leaderWorkerTemplate"] = spec["leaderWorkerTemplate"].(map[string]interface{})
+	specCopy["$patch"] = "replace"
+	objCopy["spec"] = specCopy
+	return json.Marshal(objCopy)
+}
+
+func CreateLeaderWorkerSetRevision(
+	ctx context.Context,
+	k8sClient client.Client,
+	lws *leaderworkerset.LeaderWorkerSet,
+	templateHash string) (*appsv1.ControllerRevision, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	controllerHistory := History{Client: k8sClient, context: ctx}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return nil, err
+	}
+
+	currentRevision, err := NewRevision(lws, NextRevision(revisions), templateHash)
+	if err != nil {
+		log.Error(err, "Creating new revision for lws")
+		return nil, err
+	}
+
+	return controllerHistory.CreateControllerRevision(lws, currentRevision)
+}
+
+// newRevision creates a new ControllerRevision containing a patch that reapplies the target state of LeaderWorkerSet.
```

Comment:
Just so we distinguish it from `CreateRevision` which creates it in the apiserver.

```suggestion
// newRevision instantiates a ControllerRevision object containing a patch that reapplies the target state of LeaderWorkerSet.
```